package Model;

message System {
    enum Type {
        JACK = 0;
        GUESSNUM = 1;
        SYS = 2;
    }

    message Meta {
    }

    message Request {

        enum Type {
            JOIN = 0;
            NEWUSER = 1;
        }
        message JoinGame {
            required System.Type game_type = 1;
        }
        message NewUser {
            required string name = 1;
        }

        required System.Request.Type type = 1;
        optional JoinGame join = 2;
        optional NewUser new_user = 3;
    }
    message Reply {
        message JoinGame {
        }

        optional User user = 1;
        optional JoinGame join = 2;
    }
    
}

message User {
    
    optional int id = 1;
    optional string name = 2;
    optional int32 money = 3;

}

message GuessNum {

    message Meta {
    }

    message Request {
        repeated int32 player_guess = 1;
        required Meta meta = 2;
    }

    message Reply {
        repeated int32 random_nums = 1;
        optional string result = 2;
    }
}

message BlackJack {

    message Request {
        optional string command = 1;
    }
    message Reply {
        repeated int32 random_nums = 1;
    }
}

message Request {
    optional GuessNum.Request guess_num = 1;
    optional BlackJack.Request black_jack = 2;
    optional System.Request system = 3;
    required System.Type type = 4;
    optional int32 user_id = 5 [default = 0];
}

message Reply {
    optional GuessNum.Reply guess_num = 1;
    optional BlackJack.Reply black_jack = 2;
    optional System.Reply system = 3;
    required bool status = 4;
}
